### PR TITLE
fix(term): ansi: account for some wrap edge cases

### PR DESCRIPTION
**Pull request by aymanbagabas**
Properly count escape codes, better handling of breakpoints, and only break word/breakpoint when necessary.

Fixes: https://github.com/charmbracelet/x/issues/58

### CODE REVIEW

File: exp/term/ansi/wrap.go [hunk 1]
```diff
@@ -340,23 +339,23 @@ func Wrap(s string, limit int, breakpoints string) string {
 			case runeContainsAny(r, breakpoints):
 				addSpace()
 				addWord()
-				buf.WriteByte(b[i])
-				curWidth++
+				if curWidth+1 <= limit {
```
Comment (mikelorant) [inline exp/term/ansi/wrap.go:342]: Can this be simplified as?
```go
if curWidth < limit {
```

File: exp/term/ansi/wrap.go [hunk 2]
```diff
@@ -340,23 +339,23 @@ func Wrap(s string, limit int, breakpoints string) string {
...
+					bpoint.WriteByte(b[i])
+					break
+				}
+				// If we can't fit the breakpoint in the current line, we treat
+				// it as a word character.
+				fallthrough
 			default:
-				if wordLen+1 > limit {
+				if wordLen >= limit {
+					// If the word is longer than the limit, we break it
```
Comment (mikelorant) [inline exp/term/ansi/wrap.go:351]: This comment doesn't match what they code does.

File: exp/term/ansi/wrap_test.go
Comment (mikelorant) [inline exp/term/ansi/wrap_test.go:1]: Great to see you fix the test function names. Was going to point that out earlier. 👍

File: exp/term/ansi/wrap.go [hunk 3]
```diff
@@ -132,10 +134,9 @@ func Wordwrap(s string, limit int, breakpoints string) string {
 	}
 
 	addWord := func() {
-		if word.Len() == 0 {
-			return
+		if wordLen > 0 {
```
Comment (aymanbagabas) [inline exp/term/ansi/wrap.go:137]:
```suggestion
		if curWidth > 0 {
```

File: exp/term/ansi/wrap.go [hunk 4]
```diff
@@ -247,24 +245,35 @@ func Wrap(s string, limit int, breakpoints string) string {
...
 		curWidth += space.Len()
 		buf.Write(space.Bytes())
 		space.Reset()
 	}
 
 	addWord := func() {
-		if word.Len() == 0 {
-			return
+		addBpoint()
+		if wordLen > 0 {
```
Comment (aymanbagabas) [inline exp/term/ansi/wrap.go:270]:
```suggestion
		if curWidth > 0 {
```

File: exp/term/ansi/wrap_test.go [hunk 5]
```diff
@@ -149,19 +149,57 @@ var smartWrapCases = []struct {
 	{
 		name:     "long input2",
 		input:    "Rotated keys for a-good-offensive-cheat-code-incorporated/crypto-line-operating-system.",
-		expected: "Rotated keys for a-good-offensive-cheat-code-incorporated/crypto-line-operat\ning-system.",
```
Comment (mikelorant) [inline exp/term/ansi/wrap_test.go:152]: This test had me confused on what the expected result was. This was the test case my attempted fixes kept breaking (no pun intended) on.

File: exp/term/ansi/wrap.go [hunk 6]
```diff
@@ -132,10 +134,9 @@ func Wordwrap(s string, limit int, breakpoints string) string {
 	}
 
 	addWord := func() {
-		if word.Len() == 0 {
-			return
+		if curWidth > 0 {
```
Comment (aymanbagabas) [inline exp/term/ansi/wrap.go:137]:
```suggestion
		if curWidth > 0 && word.Len() > 0 {
```

File: exp/term/ansi/wrap.go [hunk 7]
```diff
@@ -247,24 +245,35 @@ func Wrap(s string, limit int, breakpoints string) string {
...
 		curWidth += space.Len()
 		buf.Write(space.Bytes())
 		space.Reset()
 	}
 
 	addWord := func() {
-		if word.Len() == 0 {
-			return
+		addBpoint()
+		if curWidth > 0 {
```
Comment (aymanbagabas) [inline exp/term/ansi/wrap.go:270]:
```suggestion
		if curWidth > 0 && word.Len() > 0 {

```

**Comment by aymanbagabas** [comment]
@mikelorant Pushed a bunch of fixes for edge cases and support for NBSP (non-breaking space) character.

File: exp/term/ansi/wrap.go [hunk 8]
```diff
@@ -336,27 +344,32 @@ func Wrap(s string, limit int, breakpoints string) string {
...
+			case r == '-':
+				fallthrough
 			case runeContainsAny(r, breakpoints):
 				addSpace()
-				addWord()
-				buf.WriteByte(b[i])
-				curWidth++
-			default:
-				if wordLen+1 > limit {
+				if curWidth+wordLen+1 > limit {
```
Comment (mikelorant) [inline exp/term/ansi/wrap.go:352]: Consider dropping the `+1` and making it a `>=` comparison to simplify the logic?

File: exp/term/ansi/wrap.go [hunk 9]
```diff
@@ -199,6 +202,8 @@ func Wordwrap(s string, limit int, breakpoints string) string {
 			case unicode.IsSpace(r):
 				addWord()
 				space.WriteByte(b[i])
+			case r == '-':
```
Comment (mikelorant) [inline exp/term/ansi/wrap.go:205]: Would there be value in having a small helper function `IsDash` to be consistent to how it is done for `IsSpace`?

**Review by mikelorant — APPROVED** [review]

**Comment by mikelorant** [comment]
I'm not sure of the commit best practices used by the Charm team, but I generally recommend at this point to the developers I work with to squash their commits (seeing as there are now 8 commits). Your branch, so your rules 😄

**Comment by aymanbagabas** [comment]
> I'm not sure of the commit best practices used by the Charm team, but I generally recommend at this point to the developers I work with to squash their commits (seeing as there are now 8 commits). Your branch, so your rules 😄 

Yes, that was what I'm going for. I don't think the commits should be split since they're all related and address the same issues

File: exp/term/ansi/wrap.go (hunk 9 above)
Comment (aymanbagabas) [reply exp/term/ansi/wrap.go:205]: Perhaps, if we add support to other unicode dash characters 🤔

Comment (mikelorant) [reply exp/term/ansi/wrap.go:205]: Something to consider, certainly nothing to stop you merging once you do decide.

**Comment by mikelorant** [comment]
I'd say while there *might* be some minor improvements possible, the current state is worth merging. LGTM.

File: exp/term/ansi/wrap.go (hunk 8 above)
Comment (aymanbagabas) [inline exp/term/ansi/wrap.go:352]:
```suggestion
				if curWidth+wordLen >= limit {
```

File: exp/term/ansi/wrap.go (hunk 9 above)
Comment (aymanbagabas) [reply exp/term/ansi/wrap.go:205]: Thinking more about this, there are many variants of dashes defined in unicode. People who wish to use custom breakpoints can do so using the `breakpoints` parameter

**Pull request merged**